File: targets/widget/CoinGeckoService.swift
import Foundation

class CoinGeckoService {
    static let shared = CoinGeckoService()

    private let baseURL = "https://api.coingecko.com/api/v3"

    private init() {}

    enum FetchError: Error {
        case invalidURL
        case networkError(Error)
        case invalidResponse
        case decodingError(Error)
        case rateLimited
    }

    /// Helper to fetch with retry logic for rate limiting
    private func fetchWithRetry(url: URL, maxRetries: Int = 3) async throws -> Data {
        var lastError: Error?

        for attempt in 0..<maxRetries {
            do {
                let (data, response) = try await URLSession.shared.data(from: url)

                guard let httpResponse = response as? HTTPURLResponse else {
                    throw FetchError.invalidResponse
                }

                // Handle rate limiting
                if httpResponse.statusCode == 429 {
                    // Don't retry on rate limit - just fail fast and use cached data
                    print("[CoinGecko] Rate limited (429), using cached data instead")
                    throw FetchError.rateLimited
                }

                guard (200...299).contains(httpResponse.statusCode) else {
                    throw FetchError.invalidResponse
                }

                return data
            } catch {
                lastError = error
                if attempt < maxRetries - 1 {
                    let waitTime = pow(2.0, Double(attempt))
                    print("[CoinGecko] Error on attempt \(attempt + 1), waiting \(waitTime)s")
                    try await Task.sleep(nanoseconds: UInt64(waitTime * 1_000_000_000))
                }
            }
        }

        throw lastError ?? FetchError.networkError(NSError(domain: "CoinGecko", code: -1))
    }

    struct CoinGeckoToken: Codable {
        let id: String
        let symbol: String
        let name: String
        let current_price: Double
        let price_change_percentage_24h: Double?
        let image: String?
        let market_cap: Double
        let total_volume: Double
        let sparkline_in_7d: SparklineData?
    }

    struct SparklineData: Codable {
        let price: [Double]
    }

    /// Fetches top tokens by market cap from CoinGecko
    func fetchTopTokensByMarketCap(limit: Int = 10) async throws -> [Token] {
        let urlString = "\(baseURL)/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=\(limit)&page=1&sparkline=true&price_change_percentage=24h"

        guard let url = URL(string: urlString) else {
            throw FetchError.invalidURL
        }

        do {
            let data = try await fetchWithRetry(url: url)

            let decoder = JSONDecoder()
            let coins = try decoder.decode([CoinGeckoToken].self, from: data)

            return coins.map { coin in
                Token(
                    id: coin.id,
                    symbol: coin.symbol.uppercased(),
                    name: coin.name,
                    price: coin.current_price,
                    priceChange24h: coin.price_change_percentage_24h ?? 0,
                    iconUrl: coin.image,
                    sparkline: coin.sparkline_in_7d?.price
                )
            }
        } catch let error as DecodingError {
            throw FetchError.decodingError(error)
        } catch {
            throw FetchError.networkError(error)
        }
    }

    /// Fetches top tokens by trading volume from CoinGecko
    func fetchTopTokensByVolume(limit: Int = 10) async throws -> [Token] {
        let urlString = "\(baseURL)/coins/markets?vs_currency=usd&order=volume_desc&per_page=\(limit)&page=1&sparkline=true&price_change_percentage=24h"

        guard let url = URL(string: urlString) else {
            throw FetchError.invalidURL
        }

        do {
            let data = try await fetchWithRetry(url: url)

            let decoder = JSONDecoder()
            let coins = try decoder.decode([CoinGeckoToken].self, from: data)

            return coins.map { coin in
                Token(
                    id: coin.id,
                    symbol: coin.symbol.uppercased(),
                    name: coin.name,
                    price: coin.current_price,
                    priceChange24h: coin.price_change_percentage_24h ?? 0,
                    iconUrl: coin.image,
                    sparkline: coin.sparkline_in_7d?.price
                )
            }
        } catch let error as DecodingError {
            throw FetchError.decodingError(error)
        } catch {
            throw FetchError.networkError(error)
        }
    }

    /// Fetches specific tokens by their IDs (for watchlist)
    func fetchTokensByIds(_ tokenIds: [String]) async throws -> [Token] {
        if tokenIds.isEmpty {
            return []
        }

        let ids = tokenIds.joined(separator: ",")
        let urlString = "\(baseURL)/coins/markets?vs_currency=usd&ids=\(ids)&order=market_cap_desc&sparkline=true&price_change_percentage=24h"

        guard let url = URL(string: urlString) else {
            throw FetchError.invalidURL
        }

        do {
            let data = try await fetchWithRetry(url: url)

            let decoder = JSONDecoder()
            let coins = try decoder.decode([CoinGeckoToken].self, from: data)

            return coins.map { coin in
                Token(
                    id: coin.id,
                    symbol: coin.symbol.uppercased(),
                    name: coin.name,
                    price: coin.current_price,
                    priceChange24h: coin.price_change_percentage_24h ?? 0,
                    iconUrl: coin.image,
                    sparkline: coin.sparkline_in_7d?.price
                )
            }
        } catch let error as DecodingError {
            throw FetchError.decodingError(error)
        } catch {
            throw FetchError.networkError(error)
        }
    }
}

File: src/services/coinGeckoService.ts
import { Token } from '../../modules/expo-widget-bridge/src'

const COINGECKO_API_BASE = 'https://api.coingecko.com/api/v3'

// Helper to handle rate limiting - fail fast on 429
async function fetchWithRetry(url: string, maxRetries = 2): Promise<Response> {
  let lastError: Error | null = null

  for (let i = 0; i < maxRetries; i++) {
    try {
      const response = await fetch(url)

      // If rate limited, fail immediately - don't spam the API
      if (response.status === 429) {
        console.log(`[CoinGecko] Rate limited (429), aborting to avoid further rate limit hits`)
        throw new Error('CoinGecko API error: 429')
      }

      return response
    } catch (error) {
      lastError = error as Error
      // Only retry on network errors, not rate limits
      if (i < maxRetries - 1 && !(error as Error).message.includes('429')) {
        const waitTime = 1000 // Just 1 second between retries
        console.log(`[CoinGecko] Network error, waiting ${waitTime}ms before retry ${i + 1}/${maxRetries}`)
        await new Promise<void>(resolve => setTimeout(resolve, waitTime))
      } else {
        throw error
      }
    }
  }

  throw lastError || new Error('Max retries exceeded')
}

export interface CoinGeckoToken {
  id: string
  symbol: string
  name: string
  current_price: number
  price_change_percentage_24h: number
  image: string
  market_cap: number
  total_volume: number
}

/**
 * Fetches top tokens by market cap from CoinGecko
 * @param limit Number of tokens to fetch (default: 10)
 * @returns Array of tokens
 */
export async function fetchTopTokensByMarketCap(limit: number = 10): Promise<Token[]> {
  try {
    const url = `${COINGECKO_API_BASE}/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=${limit}&page=1&sparkline=false&price_change_percentage=24h`
    const response = await fetchWithRetry(url)

    if (!response.ok) {
      throw new Error(`CoinGecko API error: ${response.status}`)
    }

    const data: CoinGeckoToken[] = await response.json()

    return data.map(coin => ({
      id: coin.id,
      symbol: coin.symbol.toUpperCase(),
      name: coin.name,
      price: coin.current_price,
      priceChange24h: coin.price_change_percentage_24h || 0,
      iconUrl: coin.image,
    }))
  } catch (error) {
    console.error('[CoinGecko] Error fetching market cap data:', error)
    throw error
  }
}

/**
 * Fetches top tokens by trading volume from CoinGecko
 * @param limit Number of tokens to fetch (default: 10)
 * @returns Array of tokens
 */
export async function fetchTopTokensByVolume(limit: number = 10): Promise<Token[]> {
  try {
    const url = `${COINGECKO_API_BASE}/coins/markets?vs_currency=usd&order=volume_desc&per_page=${limit}&page=1&sparkline=false&price_change_percentage=24h`
    const response = await fetchWithRetry(url)

    if (!response.ok) {
      throw new Error(`CoinGecko API error: ${response.status}`)
    }

    const data: CoinGeckoToken[] = await response.json()

    return data.map(coin => ({
      id: coin.id,
      symbol: coin.symbol.toUpperCase(),
      name: coin.name,
      price: coin.current_price,
      priceChange24h: coin.price_change_percentage_24h || 0,
      iconUrl: coin.image,
    }))
  } catch (error) {
    console.error('[CoinGecko] Error fetching volume data:', error)
    throw error
  }
}

/**
 * Fetches specific tokens by their CoinGecko IDs (for watchlist)
 * @param tokenIds Array of CoinGecko token IDs (e.g., ['bitcoin', 'ethereum'])
 * @returns Array of tokens
 */
export async function fetchTokensByIds(tokenIds: string[]): Promise<Token[]> {
  if (tokenIds.length === 0) {
    return []
  }

  try {
    const ids = tokenIds.join(',')
    const url = `${COINGECKO_API_BASE}/coins/markets?vs_currency=usd&ids=${ids}&order=market_cap_desc&sparkline=false&price_change_percentage=24h`
    const response = await fetchWithRetry(url)

    if (!response.ok) {
      throw new Error(`CoinGecko API error: ${response.status}`)
    }

    const data: CoinGeckoToken[] = await response.json()

    return data.map(coin => ({
      id: coin.id,
      symbol: coin.symbol.toUpperCase(),
      name: coin.name,
      price: coin.current_price,
      priceChange24h: coin.price_change_percentage_24h || 0,
      iconUrl: coin.image,
    }))
  } catch (error) {
    console.error('[CoinGecko] Error fetching tokens by IDs:', error)
    throw error
  }
}
